Subject:       Changes to boot from plain dm-crypt device (v3)
Patch Author:  Orson Teodoro <orsonteodoro@hotmail.com>
Patch Status:  Complete on 4.2.3
Date:          Mon Jul 19 12:57:35 PM PDT 2021 (Unix time: 1626724655)

v3 - reduce conditional, added USE_CRYPTSETUP=1, genkernel 4.2.3 compatibility
v2 - genkernel 4.0.10 compatibility
v1 - initial release

diff -urp genkernel-4.2.3.orig/defaults/initrd.scripts genkernel-4.2.3/defaults/initrd.scripts
--- genkernel-4.2.3.orig/defaults/initrd.scripts	2021-07-08 14:09:46.000000000 -0700
+++ genkernel-4.2.3/defaults/initrd.scripts	2021-07-19 12:55:47.993053293 -0700
@@ -1682,7 +1682,7 @@ start_volumes() {
 			then
 				good_msg "ZFS pool ${ZFS_POOL} already imported."
 
-				if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_SWAP}" ]
+				if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_SWAP}" -o -n "${CRYPT_ROOT_PLAIN}" ]
 				then
 					good_msg "LUKS detected. Reimporting ${ZFS_POOL} ..."
 
@@ -1835,7 +1835,7 @@ openLUKS() {
 	local DEV_ERROR=0 KEY_ERROR=0 KEYDEV_ERROR=0
 	local mntkey="/mnt/key/" crypt_filter_ret=
 
-	if [ -z "${LUKS_DEVICE}" ]
+	if [ -z "${LUKS_DEVICE}" -a -z "${CRYPT_ROOT_PLAIN}" ]
 	then
 		bad_msg "'crypt_${1}' kernel command-line argument is not set!"
 		exit 1
@@ -1870,14 +1870,14 @@ openLUKS() {
 			KEYDEV_ERROR=0
 		else
 			LUKS_DEVICE=$(find_real_device "${LUKS_DEVICE}")
-			if [ -z "${LUKS_DEVICE}" ]
+			if [ -z "${LUKS_DEVICE}" -a -z "${CRYPT_ROOT_PLAIN}" ]
 			then
 				bad_msg "Failed to find LUKS device. If crypt_${1} kernel command-line argument is correct you are probably missing kernel support for your storage!" ${CRYPT_SILENT}
 				DEV_ERROR=1
 				continue
 			fi
 
-			if ! run cryptsetup isLuks ${LUKS_DEVICE}
+			if [ ! run cryptsetup isLuks ${LUKS_DEVICE} -a -z "${CRYPT_ROOT_PLAIN}" ]
 			then
 				bad_msg "The LUKS device ${LUKS_DEVICE} does not contain a LUKS header" ${CRYPT_SILENT}
 				DEV_ERROR=1
@@ -1975,7 +1975,12 @@ openLUKS() {
 					fi
 				fi
 				# At this point, keyfile or not, we're ready!
-				crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				if [ -n "${CRYPT_ROOT_PLAIN}" ]
+				then
+					crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} plainOpen ${CRYPT_ROOT_PLAIN} ${LUKS_NAME}"
+				else
+					crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				fi
 				crypt_filter_ret=$?
 
 				[ -e /dev/tty.org ] \
@@ -2344,7 +2349,7 @@ start_LUKS() {
 	[ -n "${CRYPT_ROOT_KEY}" ] && [ -z "${CRYPT_ROOT_KEYDEV}" ] \
 		&& sleep 6 && bootstrapKey "ROOT"
 
-	if [ -n "${CRYPT_ROOT}" ]
+	if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_ROOT_PLAIN}" ]
 	then
 		openLUKS "root"
 		if [ -n "${REAL_ROOT}" ]
diff -urp genkernel-4.2.3.orig/defaults/linuxrc genkernel-4.2.3/defaults/linuxrc
--- genkernel-4.2.3.orig/defaults/linuxrc	2021-07-08 14:09:46.000000000 -0700
+++ genkernel-4.2.3/defaults/linuxrc	2021-07-19 12:54:18.508779359 -0700
@@ -221,6 +221,10 @@ do
 		crypt_root_options=*)
 			CRYPT_ROOT_OPTIONS=$(echo ${CRYPT_ROOT_OPTIONS} ${x#*=} | sed -e 's/,/ /g')
 		;;
+		crypt_root_plain=*)
+			CRYPT_ROOT_PLAIN=${x#*=}
+			USE_CRYPTSETUP=1
+		;;
 		crypt_swap=*)
 			CRYPT_SWAP=${x#*=}
 			USE_CRYPTSETUP=1
diff -urp genkernel-4.2.3.orig/defaults/unlock-luks.sh genkernel-4.2.3/defaults/unlock-luks.sh
--- genkernel-4.2.3.orig/defaults/unlock-luks.sh	2021-07-08 14:09:46.000000000 -0700
+++ genkernel-4.2.3/defaults/unlock-luks.sh	2021-07-19 12:54:18.508779359 -0700
@@ -50,7 +50,7 @@ main() {
 	cryptsetup_options="$(trim "${cryptsetup_options}")"
 	eval local OPENED_LOCKFILE='"${CRYPT_'${TYPE}'_OPENED_LOCKFILE}"'
 
-	if [ -z "${LUKS_DEVICE}" ]
+	if [ -z "${LUKS_DEVICE}" -a -z "${CRYPT_ROOT_PLAIN}" ]
 	then
 		bad_msg "'crypt_${NAME}' kernel command-line argument is not set!"
 		exit 1
